File: core-ui/src/main/java/ru/surfstudio/android/core/ui/navigation/fragment/route/FragmentRoute.java
package ru.surfstudio.android.core.ui.navigation.fragment.route;

import android.support.v4.app.Fragment;

import ru.surfstudio.android.core.ui.navigation.Route;

/**
 * базовый маршрут для открытия фрагмента
 * см {@link Route}
 */
public abstract class FragmentRoute implements Route {

    protected abstract Class<? extends Fragment> getFragmentClass();

    public String getTag() {
        return getFragmentClass().getCanonicalName();
    }

    public Fragment createFragment() {
        try {
            return getFragmentClass().newInstance();
        } catch (InstantiationException | IllegalAccessException e) {
            throw new IllegalStateException(e);
        }
    }
}

File: navigation/lib-navigation/src/main/java/ru/surfstudio/android/navigation/route/fragment/FragmentRoute.kt
package ru.surfstudio.android.navigation.route.fragment

import androidx.fragment.app.Fragment
import ru.surfstudio.android.navigation.route.BaseRoute

/**
 * Route for [Fragment].
 */
open class FragmentRoute : BaseRoute<Fragment>() {

    /**
     * Creates Fragment with [getScreenClass] or [getScreenClassPath] class.
     */
    open fun createFragment(): Fragment {
        return requireScreenClass().newInstance()
                .apply { arguments = prepareDataWithId() }
    }
}